### ISSUE TITLE
Question about target_ids in finetune.py

### ISSUE THREAD

**Issue by syspider** [QwenLM/Qwen2#792]
![image](https://github.com/user-attachments/assets/a7b8ea1d-b6ab-4dac-aa35-b0bfd228aa6d)

Normally system message and user's instruction are not calculated into loss, which can be done by some process like replacing the corresponding ids to ignore_token_id in target_ids. 

But in the data-processing code of finetune.py, the target_ids is simply cloned by input_ids and put the ignore_token_id to pad_token_id, then nothing. Doesn't it mean that in finetune period, the system message and user' instruction (maybe multiple turns) are all involved into the loss calculation?

Why is that, is it reasonable?

**Comment by jklj077** [comment]
Hi, you are right and with that example code, the system message and user' instruction are used to calculate the loss and are learned by the model. Please refer to our previous response on this matter at <https://github.com/QwenLM/Qwen2/issues/343#issuecomment-2084495653>.

**Comment by syspider** [comment]
> Hi, you are right and with that example code, the system message and user' instruction are used to calculate the loss and are learned by the model. Please refer to our previous response on this matter at [#343 (comment)](https://github.com/QwenLM/Qwen2/issues/343#issuecomment-2084495653).

Thanks for the info! 

It is totally okay that the finetune.py serves as foundational code for anyone interested in Qwen2, just like you mentioned in issue #343 . However, to be honest, an example code should be run in bug-free. And in some way, involving the system message and user's instructions into loss calculation is actually a bug in finetuning period, which you said it was settled like that for customization. It may be more appropriate that containing full process code in example, or leaving a note in somewhere noticeable for the warning of incomplete implementation.

**Comment by jklj077** [comment]
Hi, I agree with on that we could add some notes to the code, which we will. 

The thing is that most of time we find the finetuning works fine with learning on system message (this is a must sometimes) and user messages: <https://github.com/QwenLM/Qwen2/issues/81#issuecomment-1962860429>. Certain users has also argued should or should not <|im_start|> be learned or should or should not the role be learned. One may also want to mask a certain part of the assistant message. As in the referenced issue, the performance will depend on one's expriments.

**Comment by syspider** [comment]
> Hi, I agree with on that we could add some notes to the code, which we will.
> 
> The thing is that most of time we find the finetuning works fine with learning on system message (this is a must sometimes) and user messages: [#81 (comment)](https://github.com/QwenLM/Qwen2/issues/81#issuecomment-1962860429). Certain users has also argued should or should not <|im_start|> be learned or should or should not the role be learned. One may also want to mask a certain part of the assistant message. As in the referenced issue, the performance will depend on one's expriments.

Thank your for the reply,

I think whether learning on system message (possibly including user instructions, but hardly seen before) depends on the model we choose to continually train. Just like [JustinLin610](https://github.com/JustinLin610) commented in #81 , the loss would be much higher if learning on system msg and user msgs with Chat model, because it have been done the masking strategy before and would change the inner formed pattern. That means if we use Chat model for continual training, it would be better to take masking strategy unless the task is something like role playing, function calling or etc. 

So for the example code in Qwen2 (which serves both Chat and Pretrained model), it would be much better to remind people of carefully dealing with the target masking strategy. Different masking strategy may lead to unexpected result.

Have a nice day!
